conda_smithy/feedstock_outputs.py

```diff
@@ -0,0 +1,250 @@
+
+
+def _get_ac_api():
+    """wrap this a function so we can more easily mock it when testing"""
+    from .ci_register import anaconda_token
+
+    return get_server_api(token=anaconda_token)
+
+
+def copy_feedstock_outputs(
+    outputs, staging_conda_channel, production_conda_channel
+):
+    """Copy outputs from one chanel to another.
+
+    Parameters
+    ----------
+    outputs : dict
+        A dictionary mapping each full qualified output in the conda index
+        to a hash ("md5"), its name ("name"), and version ("version").
+    staging_conda_channel : str
+        The channel that currently holds the output.
+    production_conda_channel : str
+        The channel to which to copy the output.
+
+    Returns
+    -------
+    copied : dict
+        A dict keyed on the output name with True if the copy worked and False
+        otherwise.
+    """
+    ac = _get_ac_api()
+
+    copied = {o: False for o in outputs}
+
+    for out_name, out in outputs.items():
+        try:
+            ac.copy(
+                staging_conda_channel,
+                out["name"],
+                out["version"],
+                basename=urllib.parse.quote(out_name, safe=""),
+                to_owner=production_conda_channel,
+                from_label="main",
+                to_label="main",
+            )
+            copied[out_name] = True
+        except BinstarError:
+            pass
+    return copied
+
+
+def validate_feedstock_outputs(
+    user,
+    project,
+    outputs,
+    feedstock_token,
+    conda_channel,
+    output_repo,
+    token_repo,
+    register=True,
+):
+    """Validate feedstock outputs.
+
+    Parameters
+    ----------
+    user : str
+        The GitHub user or org.
+    project : str
+        The name of the feedstock.
+    outputs : dict
+        A dictionary mapping each full qualified output in the conda index
+        to a hash ("md5"), its name ("name"), and version ("version").
+    feedstock_token : str
+        The secret token used to validate that this feedstock is who it says
+        it is.
+    conda_channel : str
+        The current conda channel with the outputs. This parameter is used to check the
+        md5 checksums from the feedstock against the ones on the channel.
+    output_repo : str
+        The GitHub repo with the feedstock output registry. This parameter is used to validate
+        that the feedstock is allowed to push to a given set of outputs.
+    token_repo : str
+        The GitHub repo with the hashed secret tokens. This repo is used as the
+        source of truth for the feedstock tokens.
+    register : bool
+        If True, attempt to register any outputs that do not exist by pushing
+        the proper json blob to `output_repo`. Default is True.
+
+    Returns
+    -------
+    valid : dict
+        A dict keyed on the keys in `outputs` with values True in the output
+        is valid and False otherwise.
+    """
+    valid = {o: False for o in outputs}
```

Comment:
Would a defaultdict be better?

Reply:
I think that that data structure is useful when you don't know the keyspace. Here we know exactly what we are outputting and so the code above struck me as clearer. Plus any python n00b will understand that snippet but they may have to look up defaultdict.

Reply:
Alright, that makes sense